File: solutions/104.maximum-depth-of-binary-tree.py
#
# @lc app=leetcode id=104 lang=python3
#
# [104] Maximum Depth of Binary Tree
#

# @lc code=start
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
from typing import Deque


class Solution:
    # Recursive solution
    def maxDepth_rec(self, root: Optional[TreeNode]) -> int:
        if not root:
            return 0

        return max(self.maxDepth(root.left), self.maxDepth(root.right)) + 1

    # Breadth-First Search (BFS) solution
    def maxDepth_bfs(self, root: Optional[TreeNode]) -> int:
        queue = Deque()
        if root:
            queue.append(root)
        depth = 0

        while queue:
            for i in range(len(queue)):
                node = queue.popleft()
                if node.left:
                    queue.append(node.left)
                if node.right:
                    queue.append(node.right)
            depth += 1

        return depth

    # Iterative Depth-First Search (DFS)
    def maxDepth(self, root: Optional[TreeNode]) -> int:
        stack = [[root, 1]]
        max_depth = 0

        while stack:
            node, depth = stack.pop()
            if node:
                max_depth = max(max_depth, depth)
                stack.append([node.right, depth + 1])
                stack.append([node.left, depth + 1])

        return max_depth

# @lc code=end

File: solutions/104.maximum-depth-of-binary-tree.java
/*
 * @lc app=leetcode id=104 lang=java
 *
 * [104] Maximum Depth of Binary Tree
 */

// @lc code=start

import java.util.LinkedList;
import java.util.Queue;
import java.util.Stack;

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    // Recursive solution
    public int maxDepth_rec(TreeNode root) {
        if (root == null) return 0;

        return Math.max(maxDepth(root.left), maxDepth(root.right)) + 1;
    }

    // Breadth-First Search (BFS) solution
    public int maxDepth_bfs(TreeNode root) {
        Queue<TreeNode> queue = new LinkedList<>();
        if (root != null) {
            queue.add(root);
        }

        int level = 0;
        while (!queue.isEmpty()) {
            int size = queue.size();
            for(int i = 0; i < size; i++) {
                TreeNode node = queue.poll();
                if (node.left != null) {
                    queue.add(node.left);
                }
                if (node.right != null) {
                    queue.add(node.right);
                }
            }
            level++;
        }
        return level;
    }

    // Iterative Depth-First Search (DFS)
    public int maxDepth(TreeNode root) {
        Stack<NodeWithDepth> stack = new Stack<>();
        stack.push(new NodeWithDepth(root, 1));
        int maxDepth = 0;

        while (!stack.isEmpty()) {
            NodeWithDepth node = stack.pop();
            if (node.node != null) {
                maxDepth = Math.max(maxDepth, node.depth);
                stack.push(new NodeWithDepth(node.node.right, node.depth + 1));
                stack.push(new NodeWithDepth(node.node.left, node.depth + 1));
            }
        }

        return maxDepth;
    }

    class NodeWithDepth {
        TreeNode node;
        int depth;

        public NodeWithDepth(TreeNode node, int depth) {
            this.node = node;
            this.depth = depth;
        }
    }
}
// @lc code=end

